### PR TITLE
Support Jina templates in requirements files

### PR DESCRIPTION
**Pull request by jamescurtin**
Closes https://github.com/dependabot/feedback/issues/769.

When parsing python requirements files, ignore Jinja delimiters (e.g. `{{ }}`, `{% %}`, etc.). This provides the ability to update dependencies that are part of a [cookiecutter](https://github.com/cookiecutter/cookiecutter) project.

### CODE REVIEW

File: python/spec/dependabot/python/file_parser_spec.rb [hunk 1]
```diff
@@ -65,6 +65,29 @@
       its(:length) { is_expected.to eq(3) }
     end
 
+    context "with jinja templates" do
+      let(:requirements_fixture_name) { "jinja_requirements.txt" }
+      describe "the first dependency" do
+        subject(:dependency) { dependencies.first }
+
+        it "has the right details" do
+          dependency = dependencies.first
```
Comment (feelepxyz) [inline python/spec/dependabot/python/file_parser_spec.rb]: I believe you can just access the subject `dependency` without assigning here
```suggestion
```

File: python/spec/dependabot/python/file_parser_spec.rb [hunk 2]
```diff
@@ -65,6 +65,29 @@
       its(:length) { is_expected.to eq(3) }
     end
 
+    context "with jinja templates" do
+      let(:requirements_fixture_name) { "jinja_requirements.txt" }
```
Comment (feelepxyz) [inline python/spec/dependabot/python/file_parser_spec.rb]: Nit, usually add a newline between `let`s and `describe/context`
```suggestion
      let(:requirements_fixture_name) { "jinja_requirements.txt" }

```

File: python/spec/dependabot/python/file_parser_spec.rb [hunk 3]
```diff
@@ -65,6 +65,29 @@
...
+      describe "the first dependency" do
+        subject(:dependency) { dependencies.first }
+
+        it "has the right details" do
+          dependency = dependencies.first
+          expect(dependency).to be_a(Dependabot::Dependency)
```
Comment (feelepxyz) [inline python/spec/dependabot/python/file_parser_spec.rb]: Don't think we need to check the type here if you're checking the `name` `version` attributes
```suggestion
```

File: python/spec/dependabot/python/file_parser_spec.rb [hunk 4]
```diff
@@ -65,6 +65,29 @@
...
+          expect(dependency.name).to eq("psycopg2")
```
Comment (feelepxyz) [inline python/spec/dependabot/python/file_parser_spec.rb]: Another way of writing this with the subject you have is replace the `it` block with this:

```
its(:name) { is_expected.to eq("psycopg2") }
its(:version) { is_expected.to eq("2.6.1") }
```

File: python/spec/dependabot/python/file_parser_spec.rb [hunk 5]
```diff
@@ -65,6 +65,29 @@
...
+          expect(dependency.version).to eq("2.6.1")
+        end
+      end
+
+      describe "the second dependency" do
+        subject(:dependency) { dependencies[1] }
```
Comment (feelepxyz) [inline python/spec/dependabot/python/file_parser_spec.rb]:
```suggestion
        subject(:dependency) { dependencies.last }

```

File: python/spec/dependabot/python/requirement_parser_spec.rb [hunk 6]
```diff
@@ -52,6 +52,11 @@ def parse(line)
       it { is_expected.to be_nil }
     end
 
+    context "with a Jinja template" do
+      let(:line) { "{{ cookiecutter.package_name }}" }
+      it { is_expected.to be_nil }
```
Comment (feelepxyz) [inline python/spec/dependabot/python/requirement_parser_spec.rb]:
```suggestion
      it { is_expected.to be_nil }

```

File: python/spec/dependabot/python/requirement_parser_spec.rb [hunk 7]
```diff
@@ -52,6 +52,11 @@ def parse(line)
       it { is_expected.to be_nil }
     end
 
+    context "with a Jinja template" do
+      let(:line) { "{{ cookiecutter.package_name }}" }
```
Comment (feelepxyz) [inline python/spec/dependabot/python/requirement_parser_spec.rb]:
```suggestion
      let(:line) { "{{ cookiecutter.package_name }}" }

```

File: python/spec/dependabot/python/requirement_parser_spec.rb [hunk 8]
```diff
@@ -79,6 +84,17 @@ def parse(line)
         end
       end
 
+      context "with an optional Jinja dependency" do
+        let(:line) do
+          "{% if cookiecutter.include_package == 'y' %} luigi==0.1.0 "\
+          "{% endif %}"
+        end
```
Comment (feelepxyz) [inline python/spec/dependabot/python/requirement_parser_spec.rb]:
```suggestion
        end

```

File: python/helpers/lib/parser.py [hunk 9]
```diff
@@ -2,17 +2,28 @@
...
+
 def parse_requirements(directory):
     # Parse the requirements.txt
     requirement_packages = []
+    parser_options = optparse.Values(
+            {
+                "skip_requirements_regex": JINJA_DELIMITER_IGNORE_REGEX,
+                "isolated_mode": False,
+                "format_control": FormatControl(),
+            }
```
Comment (feelepxyz) [inline python/helpers/lib/parser.py]: Neat! What does `isolated_mode` and `format_control` do here? Could you add a comment to explain these settings? Can definitely see myself forgetting in future 🤠

**Comment by feelepxyz** [comment]
@jamescurtin neat solution! Thanks for putting this together and writing the tests. Just asked for some style nits and comments, otherwise 👍

File: python/spec/dependabot/python/file_parser_spec.rb (hunk 1 above)
Comment (jamescurtin) [reply python/spec/dependabot/python/file_parser_spec.rb]: Good catch!

**Comment by jurre** [comment]
Hi @jamescurtin, just a FYI that I'm proposing we drop support for this feature in https://github.com/dependabot/dependabot-core/pull/1839/commits/2b0dc48884bf1b2a79269fdb184e479f4e35a158, as pip has dropped support for this `skip_requirements_regex` option.

I am curious how this is going to be supported by cookiecutter and if that is something that we can use to add support for this back. I couldn't find anything on their issue tracker and I'm not very familiar with the project.

**Comment by jamescurtin** [comment]
Thanks a lot for reaching out @jurre! The follow-up is greatly appreciated.

I have one suggested alternative approach but this is just off the top of my head, so the fitness of the solution should be validated:

Rather than parsing requirements with an internal `pip` submodule, you could instead do so with the [`pkg_resources` library](https://setuptools.readthedocs.io/en/latest/pkg_resources.html). That library contains a more usable API, so you could just parse all lines and handle lines that aren't requirements in an `except` block.

The downside is that the parsed requirements from `pkg_resources` have slightly different properties, so some refactoring of `parse_requirements` might be necessary.

Toy example to demonstrate:


```python
from pkg_resources import parse_requirements, RequirementParseError

SOME_REQUIREMENTS_TXT = """
{%- if cookiecutter.foo == "y" %}
psycopg2==2.6.1
{%- endif %}
gunicorn==20.0.2
{{ cookiecutter.another_dependency }}
"""
requirements = SOME_REQUIREMENTS_TXT.splitlines()
parsed_requirements = []

for requirement in requirements:
    try:
        parsed_requirement = parse_requirements(requirement)
    except RequirementParseError:
        # Perhaps some logic to check if we should swallow or re-raise the exception
        pass
    parsed_requirements.extend(parsed_requirement)
```

Output looks like:

```python
from pprint import pprint

pprint(parsed_requirements)
pprint(parsed_requirements[0].__dict__)

"""
[Requirement.parse('psycopg2==2.6.1'), Requirement.parse('gunicorn==20.0.2')]

{'_Requirement__hash': -8878764356812885244,
 'extras': (),
 'hashCmp': ('psycopg2', None, <SpecifierSet('==2.6.1')>, frozenset(), None),
 'key': 'psycopg2',
 'marker': None,
 'name': 'psycopg2',
 'project_name': 'psycopg2',
 'specifier': <SpecifierSet('==2.6.1')>,
 'specs': [('==', '2.6.1')],
 'unsafe_name': 'psycopg2',
 'url': None}
"""
```

**Comment by jurre** [comment]
Ooh, thanks a lot @jamescurtin that's an interesting option, I'll play around with this and see if it doesn't break any existing functionality!

**Comment by jurre** [comment]
I gave that approach a go @jamescurtin, however unfortunately it does not handle things like:

```
requests
-c constraints.txt
```

FYI this is what I had roughly so far, it works for simple requirement files, but I ran into too many issues and unfortunately don't currently have the bandwidth to dive into this further.

```python
def parse_requirements(directory):
    # Parse the requirements.txt
    requirement_packages = []
    requirement_files = glob.glob(os.path.join(directory, '*.txt')) \
                        + glob.glob(os.path.join(directory, '**', '*.txt'))

    pip_compile_files = glob.glob(os.path.join(directory, '*.in')) \
                        + glob.glob(os.path.join(directory, '**', '*.in'))

    def is_pinned(parsed_req):
        # type: () -> bool
        """Return whether requirement is pinned to an exact version.
        For example, some-package==1.2 is pinned; some-package>1.2 is not.
        """
        specifiers = parsed_req.specifier
        return (len(specifiers) == 1 and
                next(iter(specifiers)).operator in {'==', '==='})

    def version_from_install_req(install_req):
        if is_pinned(install_req):
            return next(iter(install_req.specifier)).version

    for reqs_file in requirement_files + pip_compile_files:
        parsed_requirements = []

        try:
            with open(reqs_file) as f:
                for line in f:
                    try:
                        parsed_req = parse_line(line)
                    except RequirementParseError:
                        pass

                    parsed_requirements.extend(parsed_req)

            for parsed_req in parsed_requirements:
                if parsed_req.url:
                    continue

                rel_path = os.path.relpath(reqs_file, directory)

                requirement_packages.append({
                    "name": parsed_req.project_name,
                    "version": version_from_install_req(parsed_req),
                    "markers": str(parsed_req.marker) or None,
                    "file": rel_path,
                    "requirement": str(parsed_req.specifier) or None,
                    "extras": sorted(list(parsed_req.extras))
                })
        except Exception as e:
            print(json.dumps({ "error": repr(e) }))
            exit(1)

    return json.dumps({ "result": requirement_packages })
```

**Comment by martinm82** [comment]
How about other Jinja2 template files? I am using cookiecutter templates that generate as well GitHub workflows that include some jinja code. 

Now that https://github.com/dependabot/dependabot-core/pull/6189 has been implemented, dependabot started to parse the cookiecutter workflow templates and result in failures:

```
Dependabot can't parse your quality.yml
Dependabot failed to update your dependencies because there was an error parsing the quality.yml found at /{{cookiecutter.repo_name}}/.github/workflows/quality.yml.

Dependabot encountered the following error:

Dependabot::DependencyFileNotParseable
[Learn more](https://docs.github.com/github/managing-security-vulnerabilities/troubleshooting-dependabot-errors)
```

Example workflow:

```
name: Quality Checks
on:
  pull_request:
    types:
      - opened
      - reopened
      - edited
      - synchronize
  workflow_dispatch:

jobs:
  conventional-commit:
    name: Conventional Commit Check
    runs-on: ubuntu-latest
    steps:
      {%- if cookiecutter.release_tool | lower == "commisery" %}
      - uses: tomtom-international/commisery-action@v2
        with:
          token: {% raw %}${{ github.token }}{%- endraw %}
      {%- else %}
      - uses: amannn/action-semantic-pull-request@01d5fd8a8ebb9aafe902c40c53f0f4744f7381eb
        env:
          GITHUB_TOKEN: "{% raw %}${{ secrets.GITHUB_TOKEN }}{%- endraw %}"
      {%- endif %}
```

Is this use case already covered by dependabot?

**Comment by jeffwidman** [comment]
@martinm82 this is an old PR, you'll need to open a new issue.

**Pull request merged**